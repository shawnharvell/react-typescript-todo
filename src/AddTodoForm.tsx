import React, { useState, ChangeEvent, KeyboardEvent, useEffect, useRef } from "react";
import { AddTodo } from "./types";

export interface AddTodoFormProps {
  addTodo: AddTodo;
}

export const AddTodoForm: React.FC<AddTodoFormProps> = ({ addTodo }) => {
  const [newTodo, setNewTodo] = useState<string>("");
  const inputRef = useRef<HTMLInputElement>(null);

  const handleChange = (e: ChangeEvent<HTMLInputElement>) => {
    setNewTodo(e.target.value);
  };

  const handleEnter = (e: KeyboardEvent) => {
    if (e.key === "Enter") {
      onAddClick();
    } else if (e.key === "Escape") {
      setNewTodo("");
    }
  };

  const onAddClick = () => {
    addTodo(newTodo);
    setNewTodo("");
    inputRef.current?.focus();
  };

  useEffect(() => {
    inputRef.current?.focus();
  }, []);

  return (
    <div className="add-form-container">
      <div className="textfield">
        <input
          ref={inputRef}
          type="text"
          value={newTodo}
          onKeyDown={handleEnter}
          onChange={handleChange}
          placeholder="I need to do..."
          data-testid="new-todo-textbox"
        />
      </div>
      <div className="button">
        <button onClick={onAddClick}>Add</button>
      </div>
    </div>
  );
};
