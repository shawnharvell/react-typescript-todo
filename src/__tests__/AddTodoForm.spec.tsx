import React from "react";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import "@testing-library/jest-dom";

import { AddTodoForm, AddTodoFormProps } from "../AddTodoForm";

const inputProps: AddTodoFormProps = {
  addTodo: jest.fn(),
};

describe("AddTodoForm", () => {
  it("snapshot test", () => {
    const { baseElement } = render(<AddTodoForm {...inputProps} />);
    expect(baseElement).toMatchSnapshot();
  });

  it("user interactions", async () => {
    render(<AddTodoForm {...inputProps} />);

    expect(screen.queryByPlaceholderText("I need to do...")).toHaveFocus();
    userEvent.type(screen.getByPlaceholderText("I need to do..."), "My New Todo");
    expect(screen.queryByDisplayValue("My New Todo")).toBeInTheDocument();
    userEvent.click(screen.getByText("Add"));
    expect(inputProps.addTodo).toHaveBeenCalledWith("My New Todo");

    expect(screen.getByPlaceholderText("I need to do...")).toHaveFocus();
    expect(screen.getByPlaceholderText("I need to do...")).toHaveValue("");

    userEvent.type(screen.getByPlaceholderText("I need to do..."), "My 2nd Todo{enter}");
    expect(inputProps.addTodo).toHaveBeenCalledWith("My 2nd Todo");

    expect(screen.getByPlaceholderText("I need to do...")).toHaveFocus();
    expect(screen.getByPlaceholderText("I need to do...")).toHaveValue("");

    userEvent.type(screen.getByPlaceholderText("I need to do..."), "My 3rd Todo{esc}");
    expect(inputProps.addTodo).not.toHaveBeenCalledWith("My 3rd Todo");
  });
});
