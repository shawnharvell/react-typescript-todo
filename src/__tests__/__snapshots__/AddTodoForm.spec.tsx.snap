// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddTodoForm snapshot test 1`] = `
<body>
  <div>
    <div
      class="add-form-container"
    >
      <div
        class="textfield"
      >
        <input
          data-testid="new-todo-textbox"
          placeholder="I need to do..."
          type="text"
          value=""
        />
      </div>
      <div
        class="button"
      >
        <button>
          Add
        </button>
      </div>
    </div>
  </div>
</body>
`;
